### ISSUE TITLE
[BUG] thread_pool shutdown dead lock

### ISSUE THREAD

**Issue by LazyDodo** [OpenImageIO/oiio#3851]
**Describe the bug**
 
Occasionally idiff.exe refuses to shutdown on the blender CI environment, previously openexr was to blame here but with recent developments there that specific problem appears to be solved, however a new one has sprung up in its place. This one will be a bit more annoying since i've been unable to reproduce it on my local environment sofar, best I have is a process dump from the CI environment. which shows the following stack with only the main thread still being alive. 

```
ntdll.dll!NtYieldExecution()
KERNELBASE.dll!SwitchToThread()
[Inline Frame] OpenImageIO_Util.dll!std::this_thread::yield() Line 179
[Inline Frame] OpenImageIO_Util.dll!OpenImageIO_v2_4::atomic_backoff::operator()() Line 146
OpenImageIO_Util.dll!OpenImageIO_v2_4::spin_mutex::lock() Line 219
[Inline Frame] OpenImageIO_Util.dll!std::unique_lock<OpenImageIO_v2_4::spin_mutex>::{ctor}(OpenImageIO_v2_4::spin_mutex &) Line 137
[Inline Frame] OpenImageIO_Util.dll!OpenImageIO_v2_4::pvt::ThreadsafeQueue<std::function<void __cdecl(int)> *>::pop(std::function<void __cdecl(int)> * &) Line 63
OpenImageIO_Util.dll!OpenImageIO_v2_4::thread_pool::Impl::clear_queue() Line 178
OpenImageIO_Util.dll!OpenImageIO_v2_4::thread_pool::Impl::stop(bool isWait) Line 249
[Inline Frame] OpenImageIO_Util.dll!OpenImageIO_v2_4::thread_pool::Impl::{dtor}() Line 122
OpenImageIO_Util.dll!OpenImageIO_v2_4::thread_pool::Impl::`scalar deleting destructor'(unsigned int)
OpenImageIO_Util.dll!`OpenImageIO_v2_4::default_thread_pool'::`2'::`dynamic atexit destructor for 'shared_pool''()
ucrtbase.dll!<lambda>(void)()
ucrtbase.dll!__crt_seh_guarded_call<int>::operator()<<lambda_7777bce6b2f8c936911f934f8298dc43>,<lambda>(void) &,<lambda_3883c3dff614d5e0c5f61bb1ac94921c>>()
ucrtbase.dll!_execute_onexit_table()
OpenImageIO_Util.dll!dllmain_crt_process_detach(const bool is_terminating) Line 182
OpenImageIO_Util.dll!dllmain_dispatch(HINSTANCE__ * const instance, const unsigned long reason, void * const reserved) Line 293
ntdll.dll!LdrpCallInitRoutine()
ntdll.dll!LdrShutdownProcess()
ntdll.dll!RtlExitUserProcess()
kernel32.dll!ExitProcessImplementation()
ucrtbase.dll!exit_or_terminate_process()
ucrtbase.dll!common_exit()
idiff.exe!__scrt_common_main_seh() Line 295
kernel32.dll!BaseThreadInitThunk()
ntdll.dll!RtlUserThreadStart()
````

**To Reproduce**

Oh if I only knew

**Platform information:**
 - OIIO branch/version: v2.4.11.0
 - OS: Windows 10
 - C++ compiler: MSVC 2019
 - Any non-default build flags when you build OIIO: Too many flags to list, build config can be found over here in the [blender repository](https://projects.blender.org/blender/blender/src/branch/main/build_files/build_environment/cmake/openimageio.cmake#L44)

**Comment by LazyDodo** [comment]
Managed to get a repro locally, what's happening is, all threads but the main thread are gone, and the main thread is calling `thread_pool::clear_queue()` which calls `this->q.pop(_f)` which tries to lock the `spin_mutex` inside `ThreadsafeQueue` which is locked. I'm a bit at a loss on how that could happen though

**Comment by LazyDodo** [comment]
Oh this is a "fun" one, first of I'm seemingly not the first one to run into this #2382 got close to figuring it out, but didn't cross the finish line.  It had a good tip though, calling `default_thread_pool()->resize(0);` at the end of idiff made the problem less frequent, but alas the problem is still there (occurring every 30-40 minutes rather than every 5 mins or so with the repro posted) 

What is happening is

1) I diff runs till completion, and exits its main function, at this point in time the default thread pool is still very much alive and well.
2) The standard windows process shut down commences which 
2a) Kills all threads except the main thread in the [worst way possible](https://web.archive.org/web/20190127180716/https://blogs.msdn.microsoft.com/oldnewthing/20070503-00/?p=27003)
2b) Unloads any shared libraries and sends a `DLL_PROCESS_DETACH` message to `OpenImageIO_Util.dll`, which triggers cleanup and causing dtors to run which causes `thread_pool::stop` to run.
3) Because 2a killed a thread at _just_ the wrong moment, just as it was holding the `spin_mutex` lock inside `ThreadsafeQueue` this lock is now eternally held by a essentially a ghost.
4) `thread_pool::stop` calls `thread_pool::clear_queue()` which calls `this->q.pop(_f)` which eternally will wait for the ghost to give up its lock, which will never happen.
5) Deadlock

Which leads to the question, why does calling `default_thread_pool()->resize(0);` only _partially_ solve the problem? The key there is when there's a decrease in threads, it will signal the threads it is time to exit, however `resize()` does not _wait_ for the threads to actually join. Which _most_ of the time gives them _just_ enough heads up to get their affairs in order before the OS kills them making it less likely for them to be holding the lock.

**Comment by lgritz** [comment]
Sounds like a plan to me, please do submit the PR that cleans all this up.

**Comment by lgritz** [comment]
@LazyDodo, please note that #3864 already mostly reverts #3805 -- it does the "atexit" stuff ONLY for Windows and only if the OpenEXR release is earlier than 3.1.8, which is when we believe the changes on the OpenEXR side fix the relevant bugs. (Do you think that's true?)

**Comment by lgritz** [comment]
There have been a number of patches and back-and-forth, and it's hard to follow. @LazyDodo, to be sure we're on the same page, can you please remind me, to your best assessment, which problems still persist

* with the OpenEXR thread pool
* with the OIIO internal thread pool

Does Blender use TBB?

OIIO has the ability to optionally use TBB, but its support is not quite 100% yet -- when enabled, all the parallel_for family of functions will use TBB, but there are still a few other uses of the internal thread pool that don't automatically use TBB when enabled. This shouldn't be hard for me to fix so that all thread pool use relies on TBB when enabled.

Also, I didn't know it until recently, but OpenEXR has for a long time had the ability to tell it to use an app's thread pool instead of its own internal one. I don't do that yet, but that's another thing we could try.

So, in short, it would be possible without a whole lot of work to make it so that all threading in both OIIO and OpenEXR could use TBB. Whether that's useful depends on (a) if we believe that TBB is likely to not have any of the same problems (I would like to think they got it right?), and (b) whether that would be an acceptable dependency from Blender's point of view.

**Comment by LazyDodo** [comment]
> There have been a number of patches and back-and-forth, and it's hard to follow. @LazyDodo, to be sure we're on the same page, can you please remind me, to your best assessment, which problems still persist
> with the OpenEXR thread pool

Still a problem when you combine OIIO + OpenEXR 3.1.7 or earlier, since #3805 's `atexit` runs _after_ the threads have been brutally shot in the face potentially holding a lock. The resizing the pool to side step the problem actually works, but due to _when_  the `atexit` callback runs It's too late to actually prevent the problem. 

not a problem for openEXR 3.1.8+

> with the OIIO internal thread pool

This we fixed in the last few days, no longer an issue in oiio master.

We have been running OpenEXR 3.1.7 + https://github.com/AcademySoftwareFoundation/openexr/commit/b18905772e applied on our CI env since may 11th, not a single stall in openexr since then so I am confident their fix is holding up. 

A preliminary fix (just `resize()` waiting for the threads to actually join, and resizing the oiio pool to 0 at the end idiff) for the oiio/idiff version of the problem has been deployed since jun 7th (so nearly a week now) and we have had not a single stall since (multiple times a day previously) so i'm fairly confident about this one as well. 

> Does Blender use TBB?

Blender uses TBB, but until Blender 3.6 OIIO was build without TBB support, we enabled it hoping it would use the TBB thread pool rather than OIIO's internal one and maybe not have this problem, but as you pointed out there's still a few uses left of the internal pool, so that didn't pan out for us.  Starting Blender 3.6 we ship OIIO with TBB support enabled nonetheless. 

>So, in short, it would be possible without a whole lot of work to make it so that all threading in both OIIO and OpenEXR could use TBB. Whether that's useful depends on (a) if we believe that TBB is likely to not have any of the same problems (I would like to think they got it right?), and (b) whether that would be an acceptable dependency from Blender's point of view.

a) No idea, given essentially the same problem has bitten us twice in a row, I'm unwilling to even guess if they got it right or not.  We'd have to test / look at their code to see how well they do there.
b) Not an issue for us, as mentioned we recently started building OIIO with TBB enabled 

Regardless the pending Blender 3.6 release is pretty much locked down at this point it will have to ship with the temporary patches in place.

**Comment by e4lam** [comment]
> OIIO has the ability to optionally use TBB, but its support is not quite 100% yet -- when enabled, all the parallel_for family of functions will use TBB, but there are still a few other uses of the internal thread pool that don't automatically use TBB when enabled. This shouldn't be hard for me to fix so that all thread pool use relies on TBB when enabled.

Please do! We would really love this!

> Also, I didn't know it until recently, but OpenEXR has for a long time had the ability to tell it to use an app's thread pool instead of its own internal one. I don't do that yet, but that's another thing we could try.

Reminder on my last failed attempt:
https://github.com/AcademySoftwareFoundation/openexr/issues/1285

> So, in short, it would be possible without a whole lot of work to make it so that all threading in both OIIO and OpenEXR could use TBB. Whether that's useful depends on (a) if we believe that TBB is likely to not have any of the same problems (I would like to think they got it right?), and (b) whether that would be an acceptable dependency from Blender's point of view.

Again, double +1 from me on supporting TBB purely from the number of threads that an application needs standpoint.

It should be mentioned though that using TBB does not solve the hang on shutdown problem. After of working with the TBB team on this for over 2 releases (ie. even after https://github.com/oneapi-src/oneTBB/pull/71 ), I gave up and made all our executables explicitly shutdown TBB prior to returning from main() or calling exit().
